registry/warming.go

```diff
@@ -0,0 +1,234 @@
+
+	if w.Logger == nil || w.ClientFactory == nil || w.Expiry == 0 || w.Writer == nil || w.Reader == nil {
+		panic("registry.Warmer fields are nil")
+	}
+
+	for {
+		select {
+		case <-stop:
+			w.Logger.Log("stopping", "true")
+			return
+		case r := <-warm:
+			w.warm(r)
+		}
+	}
+}
+
+func (w *Warmer) warm(id flux.ImageID) {
+	client, err := w.ClientFactory.ClientFor(id.Host)
+	if err != nil {
+		w.Logger.Log("err", err.Error())
+		return
+	}
+	defer client.Cancel()
+
+	username := w.Creds.credsFor(id.Host).username
+
+	// Refresh tags first
+	// Only, for example, "library/alpine" because we have the host information in the client above.
+	tags, err := client.Tags(id)
+	if err != nil {
+		if !strings.Contains(err.Error(), context.DeadlineExceeded.Error()) && !strings.Contains(err.Error(), "net/http: request canceled") {
+			w.Logger.Log("err", errors.Wrap(err, "requesting tags"))
+		}
+		return
+	}
+
+	val, err := json.Marshal(tags)
+	if err != nil {
+		w.Logger.Log("err", errors.Wrap(err, "serializing tags to store in cache"))
+		return
+	}
+
+	key, err := cache.NewTagKey(username, id)
+	if err != nil {
+		w.Logger.Log("err", errors.Wrap(err, "creating key for cache"))
+		return
+	}
+
+	err = w.Writer.SetKey(key, val)
+	if err != nil {
+		w.Logger.Log("err", errors.Wrap(err, "storing tags in cache"))
+		return
+	}
+
+	// Create a list of manifests that need updating
+	var toUpdate []flux.ImageID
+	var expired bool
+	for _, tag := range tags {
+		// See if we have the manifest already cached
+		// We don't want to re-download a manifest again.
+		i := id.WithNewTag(tag)
+		key, err := cache.NewManifestKey(username, i)
+		if err != nil {
+			w.Logger.Log("err", errors.Wrap(err, "creating key for memcache"))
+			continue
+		}
+		expiry, err := w.Reader.GetExpiration(key)
+		// If err, then we don't have it yet. Update.
+		if err == nil { // If no error, we've already got it
+			// If we're outside of the expiry buffer, skip, no need to update.
+			if !withinExpiryBuffer(expiry, refreshWhenExpiryWithin) {
+				continue
+			}
+			// If we're within the expiry buffer, we need to update quick!
+			expired = true
+		}
+		toUpdate = append(toUpdate, i)
+	}
+
+	if len(toUpdate) == 0 {
+		return
+	}
+
+	if expired {
+		w.Logger.Log("expiring", id.HostNamespaceImage())
+	}
+
+	// Now refresh the manifests for each tag (in lots of goroutines for improved performance)
+	toFetch := make(chan flux.ImageID, len(toUpdate))
+	fetched := make(chan flux.Image, len(toUpdate))
+	for i := 0; i < w.Burst; i++ {
+		go func() {
+			for i := range toFetch {
+				// Get the image from the remote
+				img, err := client.Manifest(i)
+				if err != nil {
+					if !strings.Contains(err.Error(), context.DeadlineExceeded.Error()) && !strings.Contains(err.Error(), "net/http: request canceled") {
+						w.Logger.Log("err", errors.Wrap(err, "requesting manifests"))
+					}
+				}
+				fetched <- img // Always return an image, otherwise the for loop below will never finish.
```

Comment:
Then just range over the channel below, instead of assuming it always has `cap` elements in it. You could also then `continue` early in this loop on error, remove the comment, and the `if img.ID.String() == "" continue` guard.